[code-completion] Fix assertion hit in getTypeOfReference via code completion

nathawes opened the pull request:
<!-- What's in this pull request? -->
```
enum QuxEnum : Int {
  case Qux1 = 10
  case Qux2 = 20
}

func foo(x: QuxEnum) {
  _ = x == .Qux1 || x == ./* code complete here */Qux2
}
```
Code completion is re-typechecking the assignment in `func foo` above. On the first pass the `UnresolvedMemberExpr`s corresponding to the enum elements are transformed into `DotSyntaxCallExpr`s. When it later re-typechecks via `typeCheckUnresolvedExpr` with the assignment as the parent, the type checker ends up in `getTypeOfReference` where references to `EnumElementDecl`s are assumed to not occur.  It looks like `typeCheckUnresolvedExpr` calls an overload of solve() that bypasses the SanitizeExpr ASTWalker, which cleans up ASTs for the constraint-based type checker. Adding that along this path seems to resolve the issue.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves Resolves rdar://problem/38144409.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

nathawes commented:
@swift-ci please smoke test

xedin reviewed (commented):
LGTM, Thanks!

benlangmuir reviewed (approved):
LGTM.  It's too bad you have to reconstruct the ast node but I guess we have no way to access the original one at this point.

nathawes commented:
@swift-ci please smoke test

nathawes commented:
The previous fix caused a regression for `CompletionKind::PostfixExpr` caught by a local run of the stress tester. Added the test case and changed the fix to focus on unresolved members specifically (`InferUnresolvedMemberConstraintGenerator` is only used when completing unresolved members from what I can tell).

Does this look ok @benlangmuir ?

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -3431,6 +3431,17 @@ class InferUnresolvedMemberConstraintGenerator : public ConstraintGenerator {
     return Expr->getType();
   }
 
+  Type visitDeclRefExpr(DeclRefExpr *Expr) override {
```
I think we should review what `InferUnresolvedMemberConstraintGenerator` is supposed to do and maybe instead just expect the number of callbacks we have for constraint generation...

slavapestov commented on lib/Sema/CSGen.cpp:
```diff
@@ -3431,6 +3431,17 @@ class InferUnresolvedMemberConstraintGenerator : public ConstraintGenerator {
+        return Ty;
+      return ConstraintGenerator::visitDeclRefExpr(Expr);
+    }
+    // Otherwise, create a type variable saying we know nothing about this expr.
```
Instead of having a subclass for this purpose it could be a flag for the constraint generator

Commit:
[code-completion] Use the SanitizeExpr ASTWalker along all paths to the constraint based type checker

Code completions calls typecheckUnresolvedExpression when completing unresolved members.
This calls an overload of solve() that bypasses sanitization, and without it we hit an
assertion failure in getTypeOfReference.

Resolves rdar://problem/38144409.

xedin commented:
@nathawes Yes, that looks great!

Commit:
Update fixed IDE validation-test crasher and move into IDE/crashers_2_fixed

nathawes commented:
@swift-ci please test

nathawes commented:
Failures seem unrelated
@swift-ci please test

nathawes commented:
@swift-ci smoke test OSX

nathawes commented:
@swift-ci smoke test OSX

nathawes commented:
@swift-ci please smoke test OS X
